services/background-worker/src/Wire/BackendNotificationPusher.hs

```diff
@@ -70,16 +72,31 @@ pushNotification runningFlag targetDomain (msg, envelope) = do
     recovering policy handlers $
       const $
         -- Ensure that the mvars are reset correctly.
+        -- Ensure that the mvars are reset correctly.
+        -- Ensure that the mvars are reset correctly.
+        -- takeMVar also has the nice feature of being a second layer of protection
+        -- takeMVar also has the nice feature of being a second layer of protection
         -- takeMVar also has the nice feature of being a second layer of protection
         -- against lazy thread updates in `amqp`. If this somehow gets called while
+        -- against lazy thread updates in `amqp`. If this somehow gets called while
+        -- against lazy thread updates in `amqp`. If this somehow gets called while
+        -- we are trying to cleanup workers for a shutdown, this will call will block
         -- we are trying to cleanup workers for a shutdown, this will call will block
+        -- we are trying to cleanup workers for a shutdown, this will call will block
+        -- and prevent the message from being sent out as we are tearing down resources.
+        -- and prevent the message from being sent out as we are tearing down resources.
         -- and prevent the message from being sent out as we are tearing down resources.
         -- This removes one way that a message might be delivered twice.
+        -- This removes one way that a message might be delivered twice.
+        -- This removes one way that a message might be delivered twice.
         UnliftIO.bracket_ (takeMVar runningFlag) (putMVar runningFlag ()) go
   where
     go :: AppT IO ()
-    go = case A.eitherDecode @BackendNotification (Q.msgBody msg) of
+    -- go = case A.eitherDecode @BackendNotification (Q.msgBody msg) of
+    go = case A.eitherDecode @(PayloadBundle _) (Q.msgBody msg) of
       Left e -> do
+        -- TODO(md): try to parse this as an old 'BackendNotification' to
+        -- maintain backward compatibility.
```

Comment:
you can do it like this:

```suggestion
    go = case (A.eitherDecode @(PayloadBundle _) (Q.msgBody msg),
               A.eitherDecode @BackendNotification (Q.msgBody msg)) of
      (Right legacy, _) -> ...
      (Left e, Right legacy) -> ...
      ...
```

This way you don't run the legacy parser unless bundle-parsing failed, and it's pretty concise and readable.

Reply:
Implemented it, though in a bit different way.